### PR TITLE
Make GetBlobsOptions and GetContainerBlobsOptions be Flags enums inst…

### PR DESCRIPTION
**Pull request by JoshLove-msft**
…ead of structs

As per [API review proposal](https://apiview.azurewebsites.net/Assemblies/Review/c52f059a-6d57-4d40-b7d2-a5816565a660#Azure.Storage.Blobs.Models.GetBlobsOptions)

- split into two separate flags enums - BlobTraits and BlobStates
- update the values to strip off the "Include" prefix.
- Prefix will be added directly as a parameter to the relevant methods (GetBlobs*)
- BlobTraits/BlobStates params will no longer be nullable in any APIs
- Analogous changes for GetBlobContainerOptions.

### CODE REVIEW

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs [hunk 1]
```diff
@@ -18,5 +20,34 @@ public void BlobDownloadInfo_Dispose()
...
             Assert.IsTrue(stream.IsDisposed);
         }
+
+        [Test]
+        public void GetBlobOptions_VariousFlagCombos()
+        {
+            GetBlobOptions options = GetBlobOptions.Metadata | GetBlobOptions.DeletedBlobs;
+            Assert.AreEqual(
+                "deleted,metadata",
+                string.Join(",", System.Linq.Enumerable.Select(options.AsIncludeItems(), item => Azure.Storage.Blobs.BlobRestClient.Serialization.ToString(item))));
```
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs]: Nit - add an `AssertOptions(string, options)` that wraps this up in a block scoped function at the bottom of this test method and call it several times.

File: sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs [hunk 2]
```diff
@@ -18,5 +20,34 @@ public void BlobDownloadInfo_Dispose()
...
+
+            options = GetBlobOptions.CopyOperationStatus | GetBlobOptions.Metadata | GetBlobOptions.Snapshots;
+            Assert.AreEqual(
+                "copy,metadata,snapshots",
+                string.Join(",", System.Linq.Enumerable.Select(options.AsIncludeItems(), item => Azure.Storage.Blobs.BlobRestClient.Serialization.ToString(item))));
+
+            options = GetBlobOptions.Metadata | GetBlobOptions.CopyOperationStatus | GetBlobOptions.Snapshots | GetBlobOptions.UncommittedBlobs | GetBlobOptions.DeletedBlobs;
+            Assert.AreEqual(
+                "copy,deleted,metadata,snapshots,uncommittedblobs",
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs]: Nit - This ordering is kind of weird.  I think metadata should be first in your `ToIncludeItems` method.  Then it's "blob details" followed by "types of blobs".

File: sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs [hunk 3]
```diff
@@ -1,8 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.Collections.Generic;
 using Azure.Storage.Blobs.Models;
 using Azure.Storage.Common.Tests.Shared;
+using Microsoft.CodeAnalysis;
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs]: Weird it didn't appear as unused in VS, but removing it doesn't cause issues.

File: sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs (hunk 2 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs]: Alphabetical order is required by the service for signing to work.. at least according to the existing comment in the AsIncludeItems method.

File: sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobContainerOptions.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+using Azure.Storage.Blobs.Models;
+
+namespace Azure.Storage.Blobs.Models
+{
+    /// <summary>
+    /// Specifies options for listing blob containers with the
+    /// <see cref="BlobServiceClient.GetBlobContainersAsync"/> operation.
+    /// </summary>
+    [Flags]
+    public enum GetBlobContainerOptions
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobContainerOptions.cs]: Ok let me know what you come up with.

File: sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/UnitTests.cs]: Oh - good catch.

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobStates.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        /// </summary>
+        Snapshots = 1,
+
+        /// <summary>
+        /// Flag specifying that blobs for which blocks have
+        /// been uploaded, but which have not been committed using
+        /// <see cref="Specialized.BlockBlobClient.CommitBlockListAsync"/> should be
+        /// included.
+        /// </summary>
+        UncommittedBlobs = 2,
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/BlobStates.cs]: We should drop `Blobs` from these names

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobTraits.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        /// <summary>
+        /// Default flag specifying that no flags are set in <see cref="BlobTraits"/>.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// Flag specifying that metadata related to any current
+        /// or previous copy operation should be included.
+        /// </summary>
+        CopyOperationStatus = 1,
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/BlobTraits.cs]: We should shorten this to `CopyStatus` - that matches the actual property name that gets returned.

File: sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobsAsyncCollection.cs [hunk 7]
```diff
@@ -43,3 +55,44 @@ internal class GetBlobsAsyncCollection : StorageCollectionEnumerator<BlobItem>
         }
     }
 }
+
+namespace Azure.Storage.Blobs
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/GetBlobsAsyncCollection.cs]: What?

File: sdk/storage/Azure.Storage.Blobs/tests/ContainerClientTests.cs [hunk 8]
```diff
@@ -1304,7 +1304,7 @@ public async Task ListBlobsFlatSegmentAsync_Prefix()
                 await SetUpContainerForListing(container);
 
                 // Act
-                IList<BlobItem> blobs = await container.GetBlobsAsync(new GetBlobsOptions { Prefix = "foo" }).ToListAsync();
+                IList<BlobItem> blobs = await container.GetBlobsAsync(states: BlobStates.None, prefix: "foo").ToListAsync();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/ContainerClientTests.cs]: Why pass `None`?

Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/ContainerClientTests.cs]: Missed that one.

File: sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs [hunk 9]
```diff
@@ -1384,9 +1416,11 @@ private void SetNameFieldsIfNull()
         /// </remarks>
         public virtual Pageable<BlobHierarchyItem> GetBlobsByHierarchy(
             string delimiter = default,
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs]: I would make `delimiter` third.

File: sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs [hunk 10]
```diff
@@ -1516,16 +1566,17 @@ private void SetNameFieldsIfNull()
                     $"{nameof(Uri)}: {Uri}\n" +
                     $"{nameof(marker)}: {marker}\n" +
                     $"{nameof(delimiter)}: {delimiter}\n" +
-                    $"{nameof(options)}: {options}");
+                    $"{nameof(traits)}: {traits}\n" +
+                    $"{nameof(states)}: {states}");
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs]: Do these have meaningful ToStrings?

**Review by tg-msft — APPROVED** [review]
